### ISSUE TITLE
i2c/spi mappings disappear from /dev on Pi Zero 2W

### ISSUE THREAD

**Issue by justsem** [pimoroni/hyperpixel4#163]
As many know by now, to get the display working under bullseye you have to remove:

dtoverlay=vc4-kms-3d from /boot/config.txt

As I was intending to use the alternate i2c bus I noticed that, when removing this overlay all SPI and I2C mappings seem to disappear from /dev.

I've been tinkering around but haven't really been able to determine a root cause or a proper solution.
The only thing I've noticed is the error thrown by i2c-bcm2835 in the kernel log.

Toying around by enabling/disabling the various i2c modules doesn't make a difference.

Another fun fact in this case is that I have another Pi zero 2W, but without the Hyperpixel display. This one - however - _does_ give me the i2c mappings even with vc4-kms-3d enabled.

So my guess is that something's going wrong in the device-tree mappings, but since that's not really my area of expertise I haven't been able to find out much as of yet.

**Things I've tried so far**
- I've tried different branches from the Hyperpixel repo (rpi3, rpi4-i2c-fix, rpi4-kms-dpi)
- I've tried enabling/disabling i2c-related kernel modules (in order to match with my other i2c-working Pi)
- I've tried various options in /boot/config.txt as was stated in various issues on GitHub

I specifically disabled/enabled the i2c_gpio and i2c_algo_bit modules as they weren't loaded on my pi that _does_ work. That didn't seem to make any difference

**Environment**
Debian Bullseye on a Pi Zero 2 W.
Hyperpixel 4 Rectengular Touch, purchased a few weeks ago.

**Debug script output**

```

## Platform Information

Revision	: 902120
Description:	Raspbian GNU/Linux 11 (bullseye)
5.10.63-v7+

## Touchscreen logs

Rectangular: Goodix
Square: ft5

## I2C Devices and Mappings

ls: cannot access '/dev/i2c-*': No such file or directory

Error: I2C bus name doesn't match any bus present!
Usage: i2cdetect [-y] [-a] [-q|-r] I2CBUS [FIRST LAST]
       i2cdetect -F I2CBUS
       i2cdetect -l
  I2CBUS is an integer or an I2C bus name
  If provided, FIRST and LAST limit the probing range.

## Boot Config

dtparam=i2c_arm=on
dtparam=spi=on

dtparam=audio=on

camera_auto_detect=1

display_auto_detect=1

max_framebuffers=2

disable_overscan=1

[cm4]
otg_mode=1

[all]
#dtoverlay=vc4-fkms-v3d

[pi4]
arm_boost=1

[all]
gpu_mem=32
dtoverlay=w1-gpio
dtdebug=1
dtoverlay=hyperpixel4
enable_dpi_lcd=1
dpi_group=2
dpi_mode=87
dpi_output_format=0x7f216
overscan_left=0
overscan_right=0
overscan_top=0
overscan_bottom=0
display_default_lcd=1
display_rotate=3
hdmi_timings=480 0 10 16 59 800 0 15 113 15 0 0 0 60 0 32000000 6
```

**dmesg**
```
[    4.266791] systemd[1]: systemd 247.3-6+rpi1 running in system mode. (+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified)
[    4.277265] systemd[1]: Detected architecture arm.
[    4.315462] systemd[1]: Set hostname to <dispi>.
[    5.331148] systemd[1]: /lib/systemd/system/plymouth-start.service:16: Unit configured to use KillMode=none. This is unsafe, as it disables systemd's process lifecycle management for the service. Please update your service to use a safer KillMode=, such as 'mixed' or 'control-group'. Support for KillMode=none is deprecated and will eventually be removed.
[    5.646666] systemd[1]: Queued start job for default target Graphical Interface.
[    5.652883] random: systemd: uninitialized urandom read (16 bytes read)
[    5.660091] systemd[1]: Created slice system-getty.slice.
[    5.669013] random: systemd: uninitialized urandom read (16 bytes read)
[    5.674240] systemd[1]: Created slice system-modprobe.slice.
[    5.682444] random: systemd: uninitialized urandom read (16 bytes read)
[    5.687890] systemd[1]: Created slice system-systemd\x2dfsck.slice.
[    5.697666] systemd[1]: Created slice system-wg\x2dquick.slice.
[    5.707155] systemd[1]: Created slice User and Session Slice.
[    5.716065] systemd[1]: Started Forward Password Requests to Wall Directory Watch.
[    5.725888] systemd[1]: Set up automount Arbitrary Executable File Formats File System Automount Point.
[    5.739893] systemd[1]: Reached target Host and Network Name Lookups.
[    5.749219] systemd[1]: Reached target Slices.
[    5.758056] systemd[1]: Reached target Swap.
[    5.767665] systemd[1]: Listening on Syslog Socket.
[    5.776878] systemd[1]: Listening on fsck to fsckd communication Socket.
[    5.785641] systemd[1]: Listening on initctl Compatibility Named Pipe.
[    5.795353] systemd[1]: Listening on Journal Audit Socket.
[    5.804650] systemd[1]: Listening on Journal Socket (/dev/log).
[    5.813851] systemd[1]: Listening on Journal Socket.
[    5.826221] systemd[1]: Listening on udev Control Socket.
[    5.835389] systemd[1]: Listening on udev Kernel Socket.
[    5.844390] systemd[1]: Condition check resulted in Huge Pages File System being skipped.
[    5.853134] systemd[1]: Mounting POSIX Message Queue File System...
[    5.866628] systemd[1]: Mounting RPC Pipe File System...
[    5.881115] systemd[1]: Mounting Kernel Debug File System...
[    5.895603] systemd[1]: Mounting Kernel Trace File System...
[    5.904440] systemd[1]: Condition check resulted in Kernel Module supporting RPCSEC_GSS being skipped.
[    5.920102] systemd[1]: Starting Restore / save the current clock...
[    5.938195] systemd[1]: Starting Set the console keyboard layout...
[    5.953929] systemd[1]: Starting Create list of static device nodes for the current kernel...
[    5.970954] systemd[1]: Starting Load Kernel Module configfs...
[    5.986833] systemd[1]: Starting Load Kernel Module drm...
[    6.004339] systemd[1]: Starting Load Kernel Module fuse...
[    6.023456] systemd[1]: Started Nameserver information manager.
[    6.038246] systemd[1]: Condition check resulted in Set Up Additional Binary Formats being skipped.
[    6.052793] systemd[1]: Starting File System Check on Root Device...
[    6.064474] fuse: init (API version 7.32)
[    6.078642] systemd[1]: Starting Journal Service...
[    6.098327] systemd[1]: Starting Load Kernel Modules...
[    6.125309] systemd[1]: Starting Coldplug All udev Devices...
[    6.172905] systemd[1]: Mounted POSIX Message Queue File System.
[    6.190081] systemd[1]: Mounted RPC Pipe File System.
[    6.199448] systemd[1]: Mounted Kernel Debug File System.
[    6.209101] systemd[1]: Mounted Kernel Trace File System.
[    6.220303] systemd[1]: Finished Restore / save the current clock.
[    6.227813] i2c /dev entries driver
[    6.237327] systemd[1]: Finished Create list of static device nodes for the current kernel.
[    6.247733] systemd[1]: modprobe@configfs.service: Succeeded.
[    6.252938] systemd[1]: Finished Load Kernel Module configfs.
[    6.267316] systemd[1]: modprobe@drm.service: Succeeded.
[    6.273016] systemd[1]: Finished Load Kernel Module drm.
[    6.285021] systemd[1]: modprobe@fuse.service: Succeeded.
[    6.292669] systemd[1]: Finished Load Kernel Module fuse.
[    6.306865] systemd[1]: Finished Load Kernel Modules.
[    6.333975] systemd[1]: Mounting FUSE Control File System...
[    6.377092] systemd[1]: Mounting Kernel Configuration File System...
[    6.391690] systemd[1]: Started File System Check Daemon to report status.
[    6.406089] systemd[1]: Starting Apply Kernel Variables...
[    6.424773] systemd[1]: Started Journal Service.
[    6.713482] EXT4-fs (mmcblk0p2): re-mounted. Opts: (null)
[    6.799435] systemd-journald[132]: Received client request to flush runtime journal.
[    6.838595] systemd-journald[132]: File /var/log/journal/f672680c6cf8441aa678418a0b28b6cd/system.journal corrupted or uncleanly shut down, renaming and replacing.
[    8.153504] vc_sm_cma: module is from the staging directory, the quality is unknown, you have been warned.
[    8.156522] bcm2835_vc_sm_cma_probe: Videocore shared memory driver
[    8.156553] [vc_sm_connected_init]: start
[    8.167379] [vc_sm_connected_init]: installed successfully
[    8.174326] mc: Linux media interface: v0.10
[    8.230453] pinctrl-bcm2835 3f200000.gpio: pin gpio9 already requested by soc:fb; cannot claim for 3f204000.spi
[    8.230481] pinctrl-bcm2835 3f200000.gpio: pin-9 (3f204000.spi) status -22
[    8.230500] pinctrl-bcm2835 3f200000.gpio: could not request pin 9 (gpio9) from group gpio9  on device pinctrl-bcm2835
[    8.230516] spi-bcm2835 3f204000.spi: Error applying setting, reverse things back
[    8.230558] spi-bcm2835: probe of 3f204000.spi failed with error -22
[    8.235103] videodev: Linux video capture interface: v2.00
[    8.286586] bcm2835_mmal_vchiq: module is from the staging directory, the quality is unknown, you have been warned.
[    8.286593] bcm2835_mmal_vchiq: module is from the staging directory, the quality is unknown, you have been warned.
[    8.289635] snd_bcm2835: module is from the staging directory, the quality is unknown, you have been warned.
[    8.292197] bcm2835_audio bcm2835_audio: there is not valid maps for state default
[    8.297436] bcm2835_v4l2: module is from the staging directory, the quality is unknown, you have been warned.
[    8.300356] bcm2835_isp: module is from the staging directory, the quality is unknown, you have been warned.
[    8.315075] bcm2835_codec: module is from the staging directory, the quality is unknown, you have been warned.
[    8.334923] bcm2835-isp bcm2835-isp: Device node output[0] registered as /dev/video13
[    8.335466] bcm2835-isp bcm2835-isp: Device node capture[0] registered as /dev/video14
[    8.335866] bcm2835-isp bcm2835-isp: Device node capture[1] registered as /dev/video15
[    8.336201] bcm2835-isp bcm2835-isp: Device node stats[2] registered as /dev/video16
[    8.336238] bcm2835-isp bcm2835-isp: Register output node 0 with media controller
[    8.336280] bcm2835-isp bcm2835-isp: Register capture node 1 with media controller
[    8.336302] bcm2835-isp bcm2835-isp: Register capture node 2 with media controller
[    8.336323] bcm2835-isp bcm2835-isp: Register capture node 3 with media controller
[    8.336518] bcm2835-isp bcm2835-isp: Loaded V4L2 bcm2835-isp
[    8.350813] bcm2835-codec bcm2835-codec: Device registered as /dev/video10
[    8.350874] bcm2835-codec bcm2835-codec: Loaded V4L2 decode
[    8.354360] bcm2835-codec bcm2835-codec: Device registered as /dev/video11
[    8.354457] bcm2835-codec bcm2835-codec: Loaded V4L2 encode
[    8.368115] bcm2835-codec bcm2835-codec: Device registered as /dev/video12
[    8.368176] bcm2835-codec bcm2835-codec: Loaded V4L2 isp
[    8.394681] bcm2835-codec bcm2835-codec: Device registered as /dev/video18
[    8.394822] bcm2835-codec bcm2835-codec: Loaded V4L2 image_fx
[    8.462164] pinctrl-bcm2835 3f200000.gpio: pin gpio2 already requested by soc:fb; cannot claim for 3f804000.i2c
[    8.462192] pinctrl-bcm2835 3f200000.gpio: pin-2 (3f804000.i2c) status -22
[    8.462211] pinctrl-bcm2835 3f200000.gpio: could not request pin 2 (gpio2) from group gpio2  on device pinctrl-bcm2835
[    8.462226] i2c-bcm2835 3f804000.i2c: Error applying setting, reverse things back
[    8.462264] i2c-bcm2835: probe of 3f804000.i2c failed with error -22
[    8.791514] cfg80211: Loading compiled-in X.509 certificates for regulatory database
[    8.951842] cfg80211: Loaded X.509 cert 'sforshee: 00b28ddf47aef9cea7'
[    9.015823] cfg80211: loaded regulatory.db is malformed or signature is missing/invalid
[    9.141471] brcmfmac: F1 signature read @0x18000000=0x1542a9a6
[    9.151533] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac43436-sdio for chip BCM43430/2
[    9.151851] usbcore: registered new interface driver brcmfmac
[    9.328618] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac43436-sdio for chip BCM43430/2
[    9.331515] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac43436-sdio for chip BCM43430/2
[    9.359251] brcmfmac: brcmf_c_preinit_dcmds: Firmware: BCM43430/2 wl0: Oct  9 2020 14:44:32 version 9.88.4.65 (test) (f149b32@shgit)  (r679549) FWID 01-f40f3270
[   10.774611] random: crng init done
[   10.774636] random: 7 urandom warning(s) missed due to ratelimiting
[   10.874268] 8021q: 802.1Q VLAN Support v1.8
[   11.030810] uart-pl011 3f201000.serial: no DMA platform data
[   11.044778] Adding 102396k swap on /var/swap.  Priority:-2 extents:2 across:1544188k SSFS
[   11.369733] brcmfmac: brcmf_cfg80211_set_power_mgmt: power save enabled
[   11.406284] Bluetooth: Core ver 2.22
```

Any ideas, pointers or suggestions are greatly welcome!

**Comment by Gadgetoid** [comment]
Does touch still work? The pins for i2c/SPI are used by Hyperpixel's DPI bus, so they should never be enabled alongside Hyperpixel and wouldn't work anyway.

I'm not sure how or why `dtoverlay=vc4-fkms-v3d` would make a difference, but I guess it could break Hyperpixel just enough that it doesn't load and trounce the i2c/SPI interfaces.

eg:

```
[    8.230453] pinctrl-bcm2835 3f200000.gpio: pin gpio9 already requested by soc:fb; cannot claim for 3f204000.spi
[    8.230481] pinctrl-bcm2835 3f200000.gpio: pin-9 (3f204000.spi) status -22
[    8.230500] pinctrl-bcm2835 3f200000.gpio: could not request pin 9 (gpio9) from group gpio9  on device pinctrl-bcm2835
[    8.230516] spi-bcm2835 3f204000.spi: Error applying setting, reverse things back
[    8.230558] spi-bcm2835: probe of 3f204000.spi failed with error -22
```

Running something like `modprobe i2c_dev` (it's this driver that actually makes the device nodes for i2c available) might get you back the `/dev/i2c-XX` node of the i2c-gpio instance HyperPixel uses for touch.

**Comment by justSem** [comment]
Yes, touch still works.
I was actually going for the i2c-3 interface as the Hyperpixel provides a breakout for that, but even that doesn't show up.
(Also: enabling the vc4-fkms-v3d breaks the hyperpixel for me, so it's just left disabled).
So my actual aim here is to be able to access that alternate bus to enable communication with a secondary microcontroller.

**Comment by justSem** [comment]
To clarify this some more:
I first went looking into the loaded kernel modules, as this more often takes a role in 'missing' interfaces.
This revealed 4 loaded modules containing the string "i2c"

```
sem@dispi:~ $ lsmod | grep i2c
i2c_bcm2835            16384  0
i2c_gpio               16384  0
i2c_algo_bit           16384  1 i2c_gpio
i2c_dev                20480  0
```

Giving the fact that the second and the third weren't present on another Pi I tried disabling it to see what it would change. For me it didn't change anything in either device listing or display functionality, so I went looking further.

So basically I'm trying to get `i2c-3` to work as described here: [https://learn.pimoroni.com/article/getting-started-with-hyperpixel-4#using-the-alternate-i2c-interface-on-hyperpixel-4-0-for-advanced-users](https://learn.pimoroni.com/article/getting-started-with-hyperpixel-4#using-the-alternate-i2c-interface-on-hyperpixel-4-0-for-advanced-users )

**Comment by rcarmo** [comment]
I'm interested in this since I would love to get an audio SHIM working... USB is possible, but would take up more space in my case.

**Comment by Gadgetoid** [comment]
@justSem sorry for the slow response. Have you tried removing `dtoverlay=w1-gpio` since *something* is causing a pin conflict on BCM 9. I don't know how or why that would break the i2c bus getting mapped in /dev but device tree is weird and I wont pretend to fully understand it :grimacing: 

@rcarmo i2s and i2c are different but incredibly confusingly named buses. i2s conflicts with the DPI bus used for HyperPixel and thus wont work. See HyperPixel pinout: https://pinout.xyz/pinout/hyperpixel4 vs PCM/i2s pinout: https://pinout.xyz/pinout/pcm

**Comment by justSem** [comment]
@Gadgetoid Yes, that was one of my attempts to see if there was anything like that bugging it.
I even tried a completely fresh bullseye install with only the Hyperpixel stuff installed, but that yielded nothing.
